File: code-monkey/src/tools/get_dependencies_tool.py
import os
from typing import Dict, Any
from .tool import Tool
from deps.deps import DependencyGraph
from constants import get_artifacts_dir
from instrumentation import instrument


class GetDependenciesTool(Tool):
    name = "get_dependencies"
    description = "Get the dependency graph for one or more Python modules"
    input_schema = {
        "type": "object",
        "properties": {
            "module_names": {
                "type": "array",
                "items": {"type": "string"},
                "description": "List of module names to get dependencies for",
            },
        },
        "required": ["module_names"],
    }

    @instrument("handle_tool_call", attributes={ "tool": "GetDependenciesTool" })
    def handle_tool_call(self, input: Dict[str, Any]) -> Dict[str, Any] | None:
        module_names = input["module_names"]
        module_paths = [
            os.path.join(get_artifacts_dir, f"{module}.py") for module in module_names()
        ]
        dependency_graph = DependencyGraph(module_paths)
        dependencies = {
            module.name: [dep.name for dep in module.dependencies]
            for module in dependency_graph.modules.values()
        }
        return str(dependencies)

File: code-monkey-ts/tools/get_dependencies_tool.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import * as path from 'path';
import { DependencyGraph } from '../deps';
import { getArtifactsDir } from '../constants';
import { instrument } from '../instrumentation';

const schema = z.object({
    module_names: z.array(z.string()).describe("List of module names to get dependencies for"),
});

export const getDependenciesTool = tool(
    async ({ module_names }: z.infer<typeof schema>) => {
        const modulePaths = module_names.map(module => path.join(getArtifactsDir(), `${module}.ts`));
        const dependencyGraph = new DependencyGraph(modulePaths);
        const dependencies = Object.fromEntries(
            Object.entries(dependencyGraph.modules).map(([name, module]) => [
                name,
                module.dependencies.map(dep => dep.name)
            ])
        );
        return JSON.stringify(dependencies);
    },
    {
        name: "get_dependencies",
        description: "Get the dependency graph for one or more TypeScript modules",
        schema: schema,
    }
);

// Decorator for instrumentation
function instrumentDecorator(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function (...args: any[]) {
        return instrument("Tool._run", { tool: "GetDependenciesTool" }, () => {
            return originalMethod.apply(this, args);
        });
    };
    return descriptor;
}

// Apply the decorator to the tool function
instrumentDecorator(getDependenciesTool, 'handler', Object.getOwnPropertyDescriptor(getDependenciesTool, 'handler')!);